web/src/engine/osk/src/keyboard-layout/oskLayerGroup.ts

```diff
@@ -165,86 +171,97 @@ export default class OSKLayerGroup {
-      if(rowRect.top <= coord.clientY && coord.clientY < rowRect.bottom) {
+      const distance = Math.abs(proportionalCoords.y - r.spec.proportionalY);
+      if(distance <= rowRadius) {
         row = r;
         break;
-      } else {
-        const distance = rowRect.top > coord.clientY ? rowRect.top - coord.clientY : coord.clientY - rowRect.bottom;
-
-        if(distance < bestMatchDistance) {
-          bestMatchDistance = distance;
-          row = r;
-        }
+      } else if(distance < bestMatchDistance) {
+        bestMatchDistance = distance;
+        row = r;
       }
     }
-
     // Assertion:  row no longer `null`.
+    // (We already prevented the no-rows available scenario, anyway.)
 
-    // Warning: am not 100% sure that what follows is actually fully correct.
-
-    // Find minimum distance from any key
-    let closestKeyIndex = 0;
-    let dx: number;
-    let dxMax = 24;
-    let dxMin = 100000;
+    // Step 2: Find minimum distance from any key
+    // - If the coord is within a key's square, go ahead and return it.
+    let closestKey: OSKBaseKey = null;
+    let minDistance = Number.MAX_VALUE;
 
-    const x = coord.clientX;
+    for (let key of row.keys) {
+      const keySpec = key.spec;
+      if(keySpec.sp == ButtonClasses.blank || keySpec.sp == ButtonClasses.spacer) {
+        continue;
+      }
 
-    for (let k = 0; k < row.keys.length; k++) {
-      // Second-biggest, though documentation suggests this is probably right.
-      const keySquare = row.keys[k].square as HTMLElement; // gets the .kmw-key-square containing a key
-      const squareRect = keySquare.getBoundingClientRect();
+      // Max distance from the key's center to consider, horizontally.
+      const keyRadius = keySpec.proportionalWidth / 2;
+      const distanceFromCenter = Math.abs(proportionalCoords.x - keySpec.proportionalX);
 
       // Find the actual key element.
-      let childNode = keySquare.firstChild ? keySquare.firstChild as HTMLElement : keySquare;
-
-      if (childNode.className !== undefined
-        && (childNode.className.indexOf('key-hidden') >= 0
-          || childNode.className.indexOf('key-blank') >= 0)) {
-        continue;
-      }
-      const x1 = squareRect.left;
-      const x2 = squareRect.right;
-      if (x >= x1 && x <= x2) {
-        // Within the key square
-        return <KeyElement>childNode;
-      }
-      dx = x1 - x;
-      if (dx >= 0 && dx < dxMin) {
-        // To right of key
-        closestKeyIndex = k; dxMin = dx;
-      }
-      dx = x - x2;
-      if (dx >= 0 && dx < dxMin) {
-        // To left of key
-        closestKeyIndex = k; dxMin = dx;
+      if(distanceFromCenter - keyRadius <= 0) {
+        // As noted above:  if we land within a key's square, match instantly!
+        return key.btn;
+      } else {
+        const distance = distanceFromCenter - keyRadius;
+        if(distance < minDistance) {
+          minDistance = distance;
+          closestKey = key;
+        }
```

Comment:
Once the distance starts to increase again, we can break out of the loop, right?